router/src/server.rs

```diff
@@ -114,6 +115,150 @@ async fn get_model_info(info: Extension<Info>) -> Json<Info> {
+                error: "Grammar and tools are mutually exclusive".to_string(),
+                error_type: "validation".to_string(),
+            }),
+        ));
+    }
+
+    let tool_grammar = match ToolGrammar::apply(tools, tool_choice) {
+        Ok(grammar) => grammar,
+        Err(err) => {
+            metrics::counter!("tgi_request_failure", "err" => "validation").increment(1);
+            tracing::error!("{}", err);
+            return Err((
+                StatusCode::UNPROCESSABLE_ENTITY,
+                Json(ErrorResponse {
+                    error: err.to_string(),
+                    error_type: err.error_type().to_string(),
+                }),
+            ));
+        }
+    };
+
+    let tools_grammar_prompt = tool_grammar.as_ref().map(|t| {
+        (
+            GrammarType::Json(serde_json::json!(t)),
+            tool_prompt.unwrap_or_default(),
+        )
+    });
+
+    let (tools_grammar_prompt, _grammar) = response_format
+        .map(|rf| (None, Some(rf)))
+        .unwrap_or_else(|| {
+            (
+                tools_grammar_prompt.clone(),
+                tools_grammar_prompt.map(|(g, _)| g),
+            )
+        });
+
+    let inputs = match infer.apply_chat_template(messages, tools_grammar_prompt) {
+        Ok(inputs) => inputs,
+        Err(err) => {
+            metrics::counter!("tgi_request_failure", "err" => "validation").increment(1);
+            tracing::error!("{}", err);
+            return Err((
+                StatusCode::UNPROCESSABLE_ENTITY,
+                Json(ErrorResponse {
+                    error: err.to_string(),
+                    error_type: err.error_type().to_string(),
+                }),
+            ));
+        }
+    };
+
+    let generate_request = GenerateRequest {
+        inputs,
+        parameters: GenerateParameters {
+            best_of: None,
+            temperature,
+            repetition_penalty: None,
+            frequency_penalty: None,
+            top_k: None,
+            top_p: None,
+            typical_p: None,
+            do_sample: true,
+            max_new_tokens: max_tokens,
+            return_full_text: None,
+            stop: stop.unwrap_or_default(),
+            truncate: None,
+            watermark: false,
+            details: false,
+            decoder_input_details: !stream,
+            seed,
+            top_n_tokens: None,
+            grammar: _grammar,
+            adapter_id: model.as_ref().filter(|m| *m != "tgi").map(String::from),
+        },
+    };
+
+    let input = generate_request.inputs.clone();
+    let encoding = infer.tokenize(generate_request).await?;
+    if let Some(encoding) = encoding {
+        let tokens: Vec<SimpleToken> = encoding
+            .get_ids()
+            .iter()
+            .zip(encoding.get_offsets())
+            .map(|(&id, &(start, stop))| {
+                let text: String =
+                    String::from_utf8_lossy(&input.as_bytes()[start..stop]).to_string();
```

Comment:
No, `start` and `stop` are defined in utf-8 char ranges.

Reply:
oh good catch, i've updated this code and the location it was copied from to operate on chars rather than bytes